subprojects/execution/src/test/groovy/org/gradle/internal/execution/steps/CaptureStateAfterExecutionStepTest.groovy

```diff
@@ -0,0 +1,117 @@
+        Stub(ValidationFinishedContext) {
+            getInputProperties() >> ImmutableSortedMap.of()
+            getInputFileProperties() >> ImmutableSortedMap.of()
+        }
+    }
+
+    def "no state is captured if before execution state is unavailable"() {
+        def delegateDuration = Duration.ofMillis(123)
+
+        when:
+        def result = step.execute(work, context)
+        then:
+        !result.afterExecutionState.present
+        !result.reused
+        result.duration == delegateDuration
+        result.originMetadata.buildInvocationId == buildInvocationScopeId.asString()
+        result.originMetadata.executionTime >= result.duration
+
+        1 * delegate.execute(work, context) >> delegateResult
+        1 * delegateResult.duration >> delegateDuration
+        _ * context.beforeExecutionState >> Optional.empty()
+        assertNoOperation()
+        0 * _
+    }
+
+    def "no state is captured if snapshotting outputs fail"() {
+        def delegateDuration = Duration.ofMillis(123)
+        def failure = new OutputSnapshotter.OutputFileSnapshottingException("output", new IOException("Error")) {}
+
+        when:
+        def result = step.execute(work, context)
+        then:
+        !result.afterExecutionState.present
+        !result.reused
+        result.duration == delegateDuration
+        result.originMetadata.buildInvocationId == buildInvocationScopeId.asString()
+        result.originMetadata.executionTime >= result.duration
+
+        1 * delegate.execute(work, context) >> delegateResult
+        1 * delegateResult.duration >> delegateDuration
+        _ * context.beforeExecutionState >> Optional.of(Mock(BeforeExecutionState) {
+            _ * detectedOverlappingOutputs >> Optional.empty()
+        })
+        1 * outputSnapshotter.snapshotOutputs(work, _) >> { throw failure }
+        assertOperation(failure)
+        0 * _
+    }
+
+    def "non-overlapping outputs are captured"() {
+        def delegateDuration = Duration.ofMillis(123)
+        def outputSnapshots = ImmutableSortedMap.<String, FileSystemSnapshot>of("outputDir", Mock(FileSystemSnapshot))
+
+        when:
+        def result = step.execute(work, context)
+        then:
+        result.afterExecutionState.get().outputFilesProducedByWork == outputSnapshots
+        !result.reused
+        result.duration == delegateDuration
+        result.originMetadata.buildInvocationId == buildInvocationScopeId.asString()
+        result.originMetadata.executionTime >= result.duration
+
+        1 * delegate.execute(work, context) >> delegateResult
+        1 * delegateResult.duration >> delegateDuration
+        _ * context.beforeExecutionState >> Optional.of(Mock(BeforeExecutionState) {
+            _ * detectedOverlappingOutputs >> Optional.empty()
+        })
+        1 * outputSnapshotter.snapshotOutputs(work, _) >> outputSnapshots
+        assertOperation()
+        0 * _
+    }
+
+    // TODO Add tests for capturing overlapping outputs
```

Comment:
Is this todo still valid?

Reply:
We should add these at some point, though it has nothing to do with this PR.